File: solutions/go/zebra-puzzle/2/zebra_puzzle.go
// Package zebra provides functionality for solving the zebra puzzle.
package zebra
const (
	red, green, ivory, yellow, blue                         = "red", "green", "ivory", "yellow", "blue"
	englishman, spaniard, ukrainian, norwegian, japanese    = "Englishman", "Spaniard", "Ukrainian", "Norwegian", "Japanese"
	dog, snails, fox, horse, zebra                          = "dog", "snails", "fox", "horse", "zebra"
	coffee, tea, milk, orangeJuice, water                   = "coffee", "tea", "milk", "orange juice", "water"
	oldGold, kools, chesterfields, luckyStrike, parliaments = "Old Gold", "Kools", "Chesterfields", "Lucky Strike", "Parliaments"
)
var colors = []string{red, green, ivory, yellow, blue}
var residents = []string{englishman, spaniard, ukrainian, norwegian, japanese}
var pets = []string{dog, snails, fox, horse, zebra}
var beverages = []string{coffee, tea, milk, orangeJuice, water}
var cigarettes = []string{oldGold, kools, chesterfields, luckyStrike, parliaments}
// Solution represents the solution to the zebra puzzle.
type Solution struct {
	DrinksWater string
	OwnsZebra   string
}
// SolvePuzzle solves the zebra puzzle.
func SolvePuzzle() Solution {
	candidates := getPermutations([]int{0, 1, 2, 3, 4})
	for _, res := range candidates {
		if residents[res[0]] != norwegian {
			continue // Condition #10
		}
	bevLoop:
		for _, bev := range candidates {
			if beverages[bev[2]] != milk {
				continue // Condition #9
			}
			for idx, elem := range bev {
				if beverages[elem] == tea && residents[res[idx]] != ukrainian {
					continue bevLoop // Condition #5
				}
			}
		colLoop:
			for _, col := range candidates {
				for idx, elem := range col {
					if colors[elem] == red && residents[res[idx]] != englishman {
						continue colLoop // Condition #2
					} else if colors[elem] == green && beverages[bev[idx]] != coffee {
						continue colLoop // Condition #4
					} else if colors[elem] == green && (idx == 0 || colors[col[idx-1]] != ivory) {
						continue colLoop // Condition #6
					} else if colors[elem] == blue {
						if (idx == 0 && residents[res[idx+1]] != norwegian) ||
							((0 < idx && idx < 4) && residents[res[idx-1]] != norwegian && residents[res[idx+1]] != norwegian) ||
							idx == 4 && residents[res[idx-1]] != norwegian {
							continue colLoop // Condition #15
						}
					}
				}
			petLoop:
				for _, pet := range candidates {
					for idx, elem := range pet {
						if pets[elem] == dog && residents[res[idx]] != spaniard {
							continue petLoop // Condition #3
						}
					}
				cigLoop:
					for _, cig := range candidates {
						for idx, elem := range cig {
							if cigarettes[elem] == oldGold && pets[pet[idx]] != snails {
								continue cigLoop // Condition #7
							} else if cigarettes[elem] == kools && colors[col[idx]] != yellow {
								continue cigLoop // Condition #8
							} else if cigarettes[elem] == chesterfields {
								if (idx == 0 && pets[pet[idx+1]] != fox) ||
									((0 < idx && idx < 4) && pets[pet[idx-1]] != fox && pets[pet[idx+1]] != fox) ||
									idx == 4 && pets[pet[idx-1]] != fox {
									continue cigLoop // Condition #11
								}
							} else if cigarettes[elem] == kools {
								if (idx == 0 && pets[pet[idx+1]] != horse) ||
									((0 < idx && idx < 4) && pets[pet[idx-1]] != horse && pets[pet[idx+1]] != horse) ||
									idx == 4 && pets[pet[idx-1]] != horse {
									continue cigLoop // Condition #12
								}
							} else if cigarettes[elem] == luckyStrike && beverages[bev[idx]] != orangeJuice {
								continue cigLoop // Condition #13
							} else if cigarettes[elem] == parliaments && residents[res[idx]] != japanese {
								continue cigLoop // Condition #14
							}
						}
						// if we are here, it means all conditions are met.
						water, zebra := getResidentDrinkingWater(res, bev), getResidentOwningZebra(res, pet)
						return Solution{DrinksWater: residents[water], OwnsZebra: residents[zebra]}
					}
				}
			}
		}
	}
	panic("could not solve puzzle")
}
func getPermutations(input []int) [][]int {
	if len(input) == 1 {
		return [][]int{[]int{input[0]}}
	}
	result := make([][]int, 0)
	for i := 0; i < len(input); i++ {
		tmpInput := make([]int, len(input)) // Need to prevent the original underlying array gets modified
		copy(tmpInput, input)
		permu := getPermutations(append(tmpInput[0:i], tmpInput[i+1:len(input)]...))
		for j := 0; j < len(permu); j++ {
			tmpResult := []int{input[i]}
			tmpResult = append(tmpResult, permu[j]...)
			result = append(result, tmpResult)
		}
	}
	return result
}
func getResidentDrinkingWater(res, bev []int) int {
	for i := 0; i < len(bev); i++ {
		if beverages[bev[i]] == water {
			return res[i]
		}
	}
	panic("nobody drinks water")
}
func getResidentOwningZebra(res, pet []int) int {
	for i := 0; i < len(pet); i++ {
		if pets[pet[i]] == zebra {
			return res[i]
		}
	}
	panic("nobody owns zebra")
}

File: solutions/c/zebra-puzzle/2/zebra_puzzle.c
#include "zebra_puzzle.h"
#include <stdlib.h>

typedef enum { NORWEGIAN, ENGLISHMAN, SPANIARD, UKRAINIAN, JAPANESE } nationality_t;
typedef enum { RED, GREEN, IVORY, YELLOW, BLUE } color_t;
typedef enum { DOG, SNAIL, HORSE, ZEBRA, FOX } pet_t;
typedef enum { TEA, COFFEE, MILK, ORANGE_JUICE, WATER } drink_t;
typedef enum { DANCING, PAINTING, READING, FOOTBALL, CHESS } hobby_t;


static int next_to(int a, int b) {
    return abs(a - b) == 1;
}

static int find(int *arr, int n, int value) {
    for (int i = 0; i < n; i++) {
        if (arr[i] == value) {
            return i;
        }
    }
    return -1;
}

static int next_permutation(int *arr, int n) {
    // step 1: find largest i where arr[i] < arr[i+1]
    int i = -1;

    for (int a = n - 2; a >= 0; a--) {
        if (arr[a] < arr[a + 1]) {
            i = a;
            break;
        }
    }

    // if no such i, we're at the last permutation — reset and return 0
    if (i < 0) {
        for (int a = 0; a < n/2; a++) {
            int tmp = arr[a];
            arr[a] = arr[n-1-a];
            arr[n-1-a] = tmp;
        }
        return 0;
    }

    // step 2: find largest j > i where arr[j] > arr[i]
    int j = -1;
    for (int a = n - 1; a > i; a--) {
        if (arr[a] > arr[i]) {
            j = a;
            break;
        }
    }

    // step 3: swap arr[i] and arr[j]
    int tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;

    // step 4: reverse everything from i+1 to end
    for (int a = 0; a < (n - i - 1)/2; a++) {
        int tmp = arr[i + 1 + a];
        arr[i + 1 + a] = arr[n - 1 - a];
        arr[n - 1 - a] = tmp;
    }

    return 1;
}

solution_t solve_puzzle(void) {
    nationality_t nationals[5] = { NORWEGIAN, ENGLISHMAN, SPANIARD, UKRAINIAN, JAPANESE };
    color_t colors[5] = { RED, GREEN, IVORY, YELLOW, BLUE };
    pet_t pets[5] = { DOG, SNAIL, HORSE, ZEBRA, FOX };
    drink_t drinks[5] = { TEA, COFFEE, MILK, ORANGE_JUICE, WATER };
    hobby_t hobbies[5] = { DANCING, PAINTING, READING, FOOTBALL, CHESS };

    const char *names[] = {"Norwegian", "Englishman", "Spaniard", "Ukrainian", "Japanese"};

    do {
        // check nationality constraints here (clue 10)
        if (nationals[0] != NORWEGIAN) continue;
    
        do {
            // check nat+color constraints here (clues 2, 6, 15)
            // The correct check is: find the house i where nationals[i] == ENGLISHMAN, then verify colors[i] == RED.
            int eng = find((int *)nationals, 5, ENGLISHMAN);
            if (eng == -1 || colors[eng] != RED) continue;

            // Clue 6 - 6. The green house is immediately to the right of the ivory house.
            int ivory = find((int *)colors, 5, IVORY);
            int green = find((int *)colors, 5, GREEN);
            if (ivory + 1 != green) continue;

            // 15. The Norwegian lives next to the blue house.
            int blue = find((int *)colors, 5, BLUE);
            if (!next_to(0, blue)) continue;

            // Drink loop
            do {
                // 5. The Ukrainian drinks tea.
                int tea = find((int *)drinks, 5, TEA);
                if (tea == -1 || nationals[tea] != UKRAINIAN) continue;

                // 4. The person in the green house drinks coffee.
                int coffee = find((int *)drinks, 5, COFFEE);
                if (coffee == -1 || colors[coffee] != GREEN) continue;

                // 9. The person in the middle house drinks milk.
                if (drinks[2] != MILK) continue;

                // Pet loop
                do {
                    // 3. The Spaniard owns the dog.
                    int dog = find((int *)pets, 5, DOG);
                    if (dog == -1 || nationals[dog] != SPANIARD) continue;
                    
                    // Hobby Lobby loop
                    do {
                        // Clue 7: snail owner → dancing
                        // find SNAIL in pets, check hobbies[snail_house] == DANCING
                        int snail = find((int *)pets, 5, SNAIL);
                        if (snail == -1 || hobbies[snail] != DANCING) continue;

                        // Clue 8: yellow house → painting
                        // Clue 8: find YELLOW in colors, check hobbies[yellow_house] == PAINTING
                        int yellow = find((int *)colors, 5, YELLOW);
                        if (yellow == -1 || hobbies[yellow] != PAINTING) continue;

                        // Clue 11: reading next to fox
                        // find READING in hobbies, check next_to(reading_house, fox_house)
                        int reading = find((int *)hobbies, 5, READING);
                        int fox = find((int *)pets, 5, FOX);
                        if (reading == -1 || !next_to(reading, fox)) continue;

                        // Clue 12: painter next to horse
                        // find PAINTING in hobbies, check next_to(painting_house, horse_house)
                        int painting = find((int *)hobbies, 5, PAINTING);
                        int horse = find((int *)pets, 5, HORSE);
                        if (painting == -1 || !next_to(painting, horse)) continue;

                        // Clue 13: football → orange juice
                        // find FOOTBALL in hobbies, check drinks[football_house] == ORANGE_JUICE
                        int football = find((int *)hobbies, 5, FOOTBALL);
                        if (football == -1 || drinks[football] != ORANGE_JUICE) continue;

                        // Clue 14: Japanese → chess
                        // find JAPANESE in nationals, check hobbies[japanese_house] == CHESS
                        int japanese = find((int *)nationals, 5, JAPANESE);
                        if (japanese == -1 || hobbies[japanese] != CHESS) continue;

                        // return solution
                        solution_t solution;
                        // find which house has WATER.
                        int house_index = find((int *)drinks, 5, WATER);
                        solution.drinks_water = names[nationals[house_index]];

                        // find which house has ZEBRA.
                        house_index = find((int *)pets, 5, ZEBRA);
                        solution.owns_zebra = names[nationals[house_index]];

                        return solution;

                    } while (next_permutation((int *)hobbies, 5));

                } while (next_permutation((int *)pets, 5));

            } while (next_permutation((int *)drinks, 5));
       
    
        } while (next_permutation((int *)colors, 5));
    } while (next_permutation((int *)nationals, 5));

    return (solution_t){ .drinks_water = NULL, .owns_zebra = NULL };

}

